compiler/resolution/resolveFunction.cpp

```diff
@@ -516,6 +516,78 @@ void resolveSpecifiedReturnType(FnSymbol* fn) {
   }
 }
 
+static Type* findThunkResultType(FnSymbol* fn) {
+  std::vector<CallExpr*> callExprs;
+  collectCallExprs(fn->body, callExprs);
+  Type* thunkResultType = nullptr;
+  for (auto call : callExprs) {
+    if (call->isPrimitive(PRIM_THUNK_RESULT)) {
+      thunkResultType = call->typeInfo();
+      break;
+    }
+  }
+  return thunkResultType;
+}
+
+static AggregateType* buildThunkRecord(FnSymbol* fn) {
+  auto newRecord = new AggregateType(AGGREGATE_RECORD);
+  auto recordName = astr("_tr_", fn->name);
+  auto recordSym = new TypeSymbol(recordName, newRecord);
+
+  recordSym->addFlag(FLAG_THUNK_RECORD);
+  return newRecord;
+}
+
+static void setReturnAndReturnSymbolType(FnSymbol* fn, Type* retType) {
+  fn->retType = retType;
+  Symbol* retSym = fn->getReturnSymbol();
+  INT_ASSERT(retSym);
+  retSym->type = retType;
+  fn->retTag = RET_VALUE;
+}
+
+static FnSymbol* buildThunkInvokeMethod(FnSymbol* fn, Type* thunkResultType, AggregateType* newRecord) {
+
+  FnSymbol* invokeFn = new FnSymbol("invoke");
+  newRecord->thunkInvoke = invokeFn;
+
+  invokeFn->addFlag(FLAG_THUNK_INVOKE);
+  invokeFn->addFlag(FLAG_INLINE);
+  invokeFn->setMethod(true);
+
+  invokeFn->_this   = new ArgSymbol(INTENT_REF, "this", newRecord);
+  invokeFn->_this->addFlag(FLAG_ARG_THIS);
+
+  invokeFn->retType = thunkResultType;
+
+  invokeFn->insertFormalAtTail(new ArgSymbol(INTENT_BLANK, "_mt", dtMethodToken));
+  invokeFn->insertFormalAtTail(invokeFn->_this);
+  return invokeFn;
+}
+
+static void protoThunkRecord(FnSymbol* fn) {
+  auto thunkResultType = findThunkResultType(fn);
+  INT_ASSERT(thunkResultType);
+
+  // Create a record that contains the thunk state (currently: captured variables),
+  // which has a method to perform the computation.
+  auto newRecord = buildThunkRecord(fn);
+  auto invokeFn = buildThunkInvokeMethod(fn, thunkResultType, newRecord);
+
+  // Adjust the function's return type and the type of the return symbol
+  setReturnAndReturnSymbolType(fn, newRecord);
+
+  fn->defPoint->insertBefore(new DefExpr(newRecord->symbol));
+  fn->defPoint->insertBefore(new DefExpr(invokeFn));
+
+  makeRefType(newRecord);
+  normalize(invokeFn);
```

Comment:
Do you know what it would take to not call `normalize` here and create the AST in a normalized way? Whenever I call `normalize` later than the pass `normalize` things seem to get messed up.

Reply:
The existing iterator support (see `protoIteratorClass`) seems to call `normalize` and not have any issues. I could try writing normalized code, but it seems the precedent says it's fine to do this.

Reply:
I am not crazy about the precedence, but trying to write the code in a normalized way is not always easy. Not something to bother with at this point.

Resolving.